plugins/statistics.koplugin/main.lua

```diff
@@ -1862,38 +2102,75 @@ function ReaderStatistics:onPageUpdate(pageno)
     if self:isDocless() or not self.is_enabled then
         return
     end
-    self.curr_page = pageno
-    self.pages_stats[TimeVal:now().sec] = pageno
-    local mem_read_pages = 0
-    local mem_read_time = 0
-    if util.tableSize(self.pages_stats) > 1 then
-        local sorted_performance = {}
-        for time, page in pairs(self.pages_stats) do
-            table.insert(sorted_performance, time)
+
+    -- We only care about *actual* page turns ;)
+    if self.curr_page == pageno then
+        return
+    end
+
+    self.pageturn_count = self.pageturn_count + 1
+    local now_ts = TimeVal:now().sec
+
+    -- Get the previous page's last timestamp (if there is one)
+    local page_data = self.page_stat[self.curr_page]
+    -- This is a list of tuples, in insertion order, we want the last one
+    local data_tuple = page_data and page_data[#page_data]
+    -- Tuple layout is { timestamp, duration }
+    local then_ts = data_tuple and data_tuple[1]
+    -- If we don't have a previous timestamp to compare to, abort early
+    if not then_ts then
+        logger.dbg("ReaderStatistics: No timestamp for previous page", self.curr_page)
+        self.page_stat[pageno] = { { now_ts, 0 } }
+        self.curr_page = pageno
+        return
+    end
+
+    -- By now, we're sure that we actually have a tuple (and the rest of the code ensures they're sane, i.e., zero-initialized)
+    local curr_duration = data_tuple[2]
+    -- NOTE: If all goes well, given the earlier curr_page != pageno check, curr_duration should always be 0 here.
+    -- Compute the difference between now and the previous page's last timestamp
+    local diff_time = now_ts - then_ts
+    if diff_time >= self.page_min_read_sec and diff_time <= self.page_max_read_sec then
+        self.mem_read_time = self.mem_read_time + diff_time
+        -- If it's the first time we're computing a duration for this page, count it as read
+        if #page_data == 1 and curr_duration == 0 then
```

Comment:
I would resign from the #page_data == 1. Isn't it equivalent of count(DISTINCT page) in SQL?

Reply:
It does ensure that we only account for a page the first time we collect it (i.e., this is essentially #6759 adapted to the new data layout).

See the NOTE right below ;).

Reply:
In #6759 the change was to make counting pages in mem_read_pages as count(DISTINCT page), like we counted total_read_pages in database.

Now total_read_pages are counted as count(page), so here mem_read_pages should be counted the same way.

The example: you read the page twice. You will increase mem_read_time. You should also increase the mem_read_pages. Otherwise we are calculation count(DISTINCT page) in memory and count(page) in dabase.

Reply:
In #6759 the change was to make counting pages in mem_read_pages as count(DISTINCT page), like we counted total_read_pages in database.

Now total_read_pages are counted as count(page), so here mem_read_pages should be counted the same way.

The example: you read the page twice. You will increase mem_read_time. You should also increase the mem_read_pages. Otherwise we are calculation count(DISTINCT page) in memory and count(page) in dabase.

Reply:
> It isn't, it's still counted as (DISTINCT page) ;). Hence why this does the same.

Yes, you are right, sorry. Too many similar names 👍

But forget me that I'll drill it a little bit more :(

self.avg_time = (self.book_read_time + self.mem_read_time) / (self.book_read_pages + self.mem_read_pages)

we are adding book_read_pages to mem_read_pages. And book_read_pages is STATISTICS_SQL_BOOK_CAPPED_TOTALS_QUERY which is count(*) !!

Reply:
> It isn't, it's still counted as (DISTINCT page) ;). Hence why this does the same.

Yes, you are right, sorry. Too many similar names 👍

But forget me that I'll drill it a little bit more :(

self.avg_time = (self.book_read_time + self.mem_read_time) / (self.book_read_pages + self.mem_read_pages)

we are adding book_read_pages to mem_read_pages. And book_read_pages is STATISTICS_SQL_BOOK_CAPPED_TOTALS_QUERY which is count(*) !!

Reply:
My fault. That's ok!